Initial implementation of the SVD Parser plugin

officialcjunior opened the pull request:
Initial implementation of the SVD Parser plugin.

The plugin uses yxml for parsing the XML and adds the registers's name and description as flags and comments on the offset they are at.

### Videos and screenshots of the plugin in action
https://user-images.githubusercontent.com/29057155/128794524-668427fc-4e6c-4868-8306-52173368dbee.mp4

![image](https://user-images.githubusercontent.com/29057155/128794832-54182f2b-8ec5-4d90-83f2-c3f3041a3b29.png)


![image](https://user-images.githubusercontent.com/29057155/128794866-9dbc6443-d5c3-4b07-82bd-dcee5e4d3910.png)

XVilka commented:
1. I recommend to add Meson build too
2. Regarding the plugin - you could create a flag with `f <blockname>.<regname> <size> <address>` for the `peripherals` category, probably also a comment.
3. Regarding YXML usage, see more complex example https://github.com/getdnsapi/getdns/blob/develop/src/anchor.c

officialcjunior commented:
Here, I think have made it into a nearly working state.

```
$ rizin --
[0x00000000]> svd file.svd
[0x00000000]> f
0x00000009 1 GPIOA.DISOOFP
[0x00000000]>
```
Here, that flag was the last register of the [STM32F303's SVD file](https://github.com/posborne/cmsis-svd/blob/master/data/STMicro/STM32F303.svd) I had tested on. So, I think, the overall logic is pretty fine with a minor mistake somewhere along the way.

Also, while debugging by putting a fprintf at [L187](https://github.com/rizinorg/rizin-extras/pull/4/files#diff-784b07645712e153267bd55ec4e84f60a5038d9305822b974f4040b52060722fR187), I realized that the registers getting displayed differed on every time I do that, although the data was correctly read. I'm not sure whether that's the part of the plan, I was looking into that as well.

XVilka commented:
I recommend to add a test also

officialcjunior commented:
![image](https://user-images.githubusercontent.com/29057155/112524410-cbab3180-8dc5-11eb-8ba0-79c9be719735.png)

There's been some progress. I added `<description>` tag to be added as comments (Currently just `CC`-ing) and added a test file I was trying it on.

Issues:
Haven't figured out why it isn't reading the whole information, yet. I'm sure that it's got something to do with the `level`s not handled properly.

The `groupName` seems to have been not updating according to the different peripherals too. Need to fix that, as well.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,295 @@
+}
+
+static ta_iter *ta_iter_init(ta_iter *ta, const char *file) {
+	char *doc = rz_file_slurp(file, NULL);
```
I recommend to check it for NULL first.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,295 @@
+
+static ta_iter *ta_iter_init(ta_iter *ta, const char *file) {
+	char *doc = rz_file_slurp(file, NULL);
+	int doc_len = rz_file_size(file);
```
Same here, check for the error.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,295 @@
+		return NULL;
+	}
+
+	assert(r == YXML_ELEMSTART);
```
I am not sure about plugins, but since we are using RzUtil anyway, probably better to use `rz_return_if_fail()` instead of the `assert()`.

officialcjunior replied:
Sure, I'm just confused on what exactly to pass to `rz_return_if_fail()` so that the logic is preserved.

XVilka replied:
The same condition should work just fine.

officialcjunior replied:
That throws up some warnings, as the function is non-void.
```
In file included from /usr/local/include/librz/rz_vector.h:5,
                 from /usr/local/include/librz/rz_cons.h:21,
                 from /usr/local/include/librz/rz_diff.h:6,
                 from /usr/local/include/librz/rz_util.h:8,
                 from /usr/local/include/librz/rz_io.h:10,
                 from /usr/local/include/librz/rz_analysis.h:14,
                 from ../rz-svd.c:1:
../rz-svd.c: In function ‘ta_iter_next’:
/usr/local/include/librz/rz_util/rz_assert.h:100:4: warning: ‘return’ with no value, in function returning non-void [-Wreturn-type]
  100 |    return; \
      |    ^~~~~~
../rz-svd.c:156:2: note: in expansion of macro ‘rz_return_if_fail’
  156 |  rz_return_if_fail(r == YXML_ELEMSTART);
      |  ^~~~~~~~~~~~~~~~~
../rz-svd.c:62:17: note: declared here
   62 | static ta_iter *ta_iter_next(ta_iter *ta) {
      |                 ^~~~~~~~~~~~
```

XVilka replied:
Yes, because in case of the function returning something you need to use `rz_return_val_if_fail()` instead.

officialcjunior replied:
Replaced with an if/return statement. 👍🏽

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,295 @@
+#define XMLBUFSIZE 4096
+
+typedef struct ta_iter {
+	uint8_t yxml_buf[4096];
```
You could use a define here as well.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,295 @@
+
+	int level;
+	char *cur, *tmp;
+	char value[2048];
```
Probably it's better to use a define here too.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,295 @@
+	char *cur, *tmp;
+	char value[2048];
+	const char *ta_start;
+	enum { REGNAME,
```
I think it's better to define it as a type outside of the function.

officialcjunior commented:
Currently at a state where returning and adding as flags are broken, but the parse is working for all registers inside the first peripheral. 

Here's what I get when I `fprintf` it where it's getting parsed.
```
baseaddress 0x48000000
         
regname MODER 
description GPIO port mode register 
offset 0x0 
width 0x20 
```

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,309 @@
+	char baseaddress[50];
+	char bitoffset[10];
+	char bitwidth[20];
+	char description[50];
```
I think description can be much bigger than that.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,309 @@
+	{
+		.name = "Path to the SVD file",
+		.type = RZ_CMD_ARG_TYPE_FILE,
+
```
Unnecessary empty line

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,309 @@
+	cur = value;
+	value[0] = 0;
+
+	if (!ta->baseaddress[0]) {
```
Maybe better to inverse the if instead.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,309 @@
+	}
+	assert(ta->baseaddress[0]);
+
+	// find <register>
```
Probably better to split this part into a separate function

officialcjunior replied:
Done.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,309 @@
+	value[0] = 0;
+	elem_type = -1;
+
+	*ta->bitoffset = *ta->regname = *ta->bitwidth = 0;
```
I recommend to extract the following part into the separate function as well.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,309 @@
+				strcpytrunc(ta->regname, value, sizeof(ta->regname));
+				break;
+			case DESCRIPTION:
+				rz_str_replace_char(value, '\n', ' ');
```
I guess this is also relevant to https://github.com/rizinorg/rizin/issues/1356

officialcjunior replied:
I don't think so, this is just to handle descriptions like
```xml
              <description>Port x configuration bits (y =
              0..15)</description>
```

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,309 @@
+			cur = NULL;
+			switch (elem_type) {
+			case BITOFFSET:
+				strcpytrunc(ta->bitoffset, value, sizeof(ta->bitoffset));
```
I think this function is not available everywhere. Please substitute with functions from `rz_util`.

officialcjunior replied:
It's defined here in the same file, actually: https://github.com/rizinorg/rizin-extras/pull/4/files#diff-784b07645712e153267bd55ec4e84f60a5038d9305822b974f4040b52060722fR35-R41

I'll see if there's anything in `rz_util`, as well.

XVilka replied:
I think `rz_str_ncpy()` does that.

XVilka replied:
By the way, the `rz_str_ncpy()` implementation looks inefficient, might make sense to make a pull request to use `memcpy` in it.

officialcjunior replied:
Okay. Let me see :)

officialcjunior replied:
Yup! Done.

XVilka commented on rz-svd/rz-svd.c:
```diff
@@ -0,0 +1,309 @@
+		}
+		r = yxml_parse(&ta->x, *ta->ptr);
+	}
+	return ta->bitoffset && *ta->regname && *ta->baseaddress && *ta->bitwidth && *ta->description ? ta : ta_iter_next(ta);
```
Extract this logical check into a separate inline function

officialcjunior commented:
More progress.
The last register of the first `<peripheral>` gets added now with its name as flag and its comment at the right offset.
![image](https://user-images.githubusercontent.com/29057155/127342270-3863038e-aace-44c5-8feb-cc15888d03d6.png)

The reason why this happens is because only the last register is properly returned. 
Hoping to fix this next!

officialcjunior commented:
The process of parsing and everything with it is correct, it seems and the only problem is that the loop isn't broken and the values returned in the right times.

Right now, this is the place where the loop is broken:
https://github.com/officialcjunior/rizin-extras/blob/d5d064619c793b2461466d2ff391217d8c6937fc/rz-svd/rz-svd.c#L235

If I'm checking for the level as 2, only the first register is added and it is 1 or 2, the last register of the first peripheral is added.

I believe that there's someplace else where this should be handled and this place is handle the exit at the very end.

officialcjunior commented:
Right now, with the code I haven't pushed here, I am able to parse the register names and its description and add it as flags and comments for all SVD files from STM that I tested.

https://user-images.githubusercontent.com/29057155/128610093-649d4827-80d6-41da-8470-56eee1910a0d.mp4

XVilka commented:
Looks good now, but I can't play this video for some reason.
